signalr-go-server/pkg/signalr/server.go

```diff
@@ -0,0 +1,265 @@
+package signalr
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"reflect"
+	"runtime/debug"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/net/websocket"
+)
+
+// MapHub used to register a SignalR Hub with the specified ServeMux
+func MapHub(mux *http.ServeMux, path string, hubPrototype HubInterface) {
+	lifetimeManager := defaultHubLifetimeManager{}
+	defaultHubClients := defaultHubClients{
+		lifetimeManager: &lifetimeManager,
+		allCache:        allClientProxy{lifetimeManager: &lifetimeManager},
+	}
+
+	groupManager := defaultGroupManager{
+		lifetimeManager: &lifetimeManager,
+	}
+
+	hubType := reflect.TypeOf(hubPrototype)
+
+	mux.HandleFunc(fmt.Sprintf("%s/negotiate", path), negotiateHandler)
+	mux.Handle(path, websocket.Handler(func(ws *websocket.Conn) {
+		connectionID := ws.Request().URL.Query().Get("id")
+
+		if len(connectionID) == 0 {
+			// Support websocket connection without negotiate
+			connectionID = getConnectionID()
+		}
+
+		// Copy hubPrototype
+		hub := reflect.New(reflect.ValueOf(hubPrototype).Elem().Type()).Interface().(HubInterface)
```

Comment:
Why?

Reply:
To implement ClientProxy.Caller(). Each connection get an hub instance

Reply:
Yea, I'm jaded about this. If creation of the hub happens per invocation then it requires dependency injection to pass state into the hub. I don't think we should implement caller as yet. That was explicitly left out.